src/EditorFeatures/Core/GoToDefinition/AbstractGoToDefinitionSymbolService.cs

```diff
@@ -28,7 +36,43 @@ public async Task<(ISymbol, TextSpan)> GetSymbolAndBoundSpanAsync(Document docum
                          semanticInfo.DeclaredSymbol ??
                          semanticInfo.Type;
 
+            // Disabled navigation if token is a part of declaration syntax
+            // or if it is a part of partial method implementation.
+            if (symbol != null &&
+            	IsNotPartialMethodDefinitionPart(symbol) &&
+                TokenIsPartOfDeclaringSyntax(token, symbol))
+            {
+                return EmptyResult;
+            }
+
             return (FindRelatedExplicitlyDeclaredSymbol(symbol, semanticModel.Compilation), semanticInfo.Span);
         }
+
+        private async Task<SyntaxToken> GetToken(SemanticModel semanticModel, Workspace workspace, int position, CancellationToken cancellationToken)
+        {
+            var syntaxTree = semanticModel.SyntaxTree;
+            var syntaxFacts = workspace.Services.GetLanguageServices(semanticModel.Language).GetService<ISyntaxFactsService>();
+            var token = await syntaxTree.GetTouchingTokenAsync(position, syntaxFacts.IsBindableToken, cancellationToken, findInsideTrivia: true).ConfigureAwait(false);
+
+            if (token != default &&
+                token.Span.IntersectsWith(position))
+            {
+                return token;
+            }
+
+            return default;
+        }
+
+        private static bool IsNotPartialMethodDefinitionPart(ISymbol symbol)
+            => !(symbol is IMethodSymbol methodSymbol && methodSymbol.IsPartialMethodDefinitionPart());
+
+        private static bool TokenIsPartOfDeclaringSyntax(SyntaxToken token, ISymbol symbol)
+        {
+            // For partial classes symbol can contain more than one declaring syntax reference
+            // and we won't change behavior for that cases (user should be able to navigate to other class declarations).
+            return symbol.DeclaringSyntaxReferences.Length == 1 && symbol.DeclaringSyntaxReferences[0].GetSyntax().Equals(token.Parent);
```

Comment:
Agreed.  I'm also confused why this PR is working in this manner. Are we effectively only trying to say "do not go to def on contextual keywords"?  If so, we can just check if the token is a non-type keyword and prevent go-to-def on those cases.

Reply:
> Are we effectively only trying to say "do not go to def on contextual keywords"?

No, we are trying to say do not go to definition when you click on the declaring token (i.e. the target token of navigation is the token that was clicked).

Reply:
❗️ @zaytsev-victor we need to find a way to avoid the call to `GetSyntax()` here. I don't have time right now to go and look for what exactly this is; if you already know of a way then great otherwise ask and I'll make the time. 😄

Reply:
I have a fix, tomorrow I'll add more test and will update PR.